src/prommis/nanofiltration/costing/diafiltration.py

```diff
@@ -0,0 +1,713 @@
+    P_Co = round(pyo.value(m.Co_purity) * 100, 2)
+
+    membrane_length = (
+        pyo.value(m.fs.stage1.length)
+        + pyo.value(m.fs.stage2.length)
+        + pyo.value(m.fs.stage3.length)
+    ) * pyo.units.m
+
+    # the feed is the stream entering stage 3 in the last element
+    feed_flow_rate = (
+        pyo.value(m.fs.stage3.retentate_side_stream_state[0, 10].flow_vol)
+        * pyo.units.m**3
+        / pyo.units.hr
+    )
+
+    # the diafiltrate is the stream entering stage 3 in the first element
+    diafiltrate_flow_rate = (
+        pyo.value(m.fs.stage3.retentate_inlet.flow_vol[0])
+        * pyo.units.m**3
+        / pyo.units.hr
+    )
+
+    # the permeate is the stream exiting stage 3 in the last element
+    permeate_flow_rate = (
+        pyo.value(m.fs.stage3.permeate_outlet.flow_vol[0])
+        * pyo.units.m**3
+        / pyo.units.hr
+    )
+
+    return (
+        R_Li,
+        R_Co,
+        P_Li,
+        P_Co,
+        membrane_length,
+        feed_flow_rate,
+        diafiltrate_flow_rate,
+        permeate_flow_rate,
+    )
+
+
+def add_costing(
+    m,
+    membrame_length,
+    membrane_width,
+    water_flux,
+    feed_flow,
+    perm_flow,
+    inlet_pressure,
+    outlet_pressure,
+    diaf_flow,
+):
+    # creating dummy variables to store the UnitModelCostingBlocks
+    # TODO: assign costing blocks to each stage
+    m.fs.membrane = UnitModelBlock()
+    m.fs.pump = UnitModelBlock()
+
+    m.fs.costing = DiafiltrationCosting()
+    m.fs.membrane.costing = UnitModelCostingBlock(
+        flowsheet_costing_block=m.fs.costing,
+        costing_method=DiafiltrationCostingData.cost_membranes,
+        costing_method_arguments={
+            "membrane_length": membrame_length,
+            "membrane_width": membrane_width,
+            "water_flux": water_flux,
+            "vol_flow_feed": feed_flow,
+            "vol_flow_perm": perm_flow,
+        },
+    )
+
+    m.fs.pump.costing = UnitModelCostingBlock(
+        flowsheet_costing_block=m.fs.costing,
+        costing_method=DiafiltrationCostingData.cost_pump,
+        costing_method_arguments={
+            "inlet_pressure": inlet_pressure,
+            "outlet_pressure": outlet_pressure,
+            "inlet_vol_flow": diaf_flow,
+        },
+    )
+    m.fs.costing.cost_process()
+
+
+def unfix_variables(m):
+    m.fs.stage1.length.unfix()
+    m.fs.stage2.length.unfix()
+    m.fs.stage3.length.unfix()
+
+
+def add_constraints(m):
+    m.Co_recovery_constraint = pyo.Constraint(expr=m.Co_recovery >= 0.4)
+    m.Li_recovery_constraint = pyo.Constraint(expr=m.Li_recovery >= 0.95)
+
+    # m.Li_purity_constraint = pyo.Constraint(expr=m.Li_purity >= 0.8)
+    # m.Co_purity_constraint = pyo.Constraint(expr=m.Co_purity >= 0.5)
+
+
+def add_objective(m):
+    def cost_obj(m):
+        return m.fs.costing.total_annualized_cost
+
+    m.cost_objecticve = pyo.Objective(rule=cost_obj)
+
+
+def print_information(m):
```

Comment:
Not critical for now, but for these types of reporting functions it is a good idea to allow the user to specify an output stream to use.

```
import sys

def my_report(model, stream=None):
    if stream is None:
        stream = sys.stdout

    stream.write("Hello world/n")
```

This is helpful for testing as it lets you provide a `StringIO` object to capture the output of the report function and check that it is correct without having to mess around with `capsys`.